semtypes/src/main/java/io/ballerina/types/typeops/BddCommonOps.java

```diff
@@ -119,12 +119,16 @@ public static Bdd bddDiff(Bdd b1, Bdd b2) {
                         bddDiff(b1, bddUnion(((BddNode) b2).right, ((BddNode) b2).middle)));
 
             } else {
-                // This is incorrect in the AMK tutorial
-                // but correct in the Castagna paper
+                // There is an error in the Castagna paper for this formula.
+                // The union needs to be materialized here.
+                // The original formula does not work in a case like (a0|a1) - a0.
+                // Castagna confirms that the following formula is the correct one.
                 return bddCreate(((BddNode) b1).atom,
-                        bddDiff(((BddNode) b1).left, ((BddNode) b2).left),
-                        bddDiff(((BddNode) b1).middle, ((BddNode) b2).middle),
-                        bddDiff(((BddNode) b1).right, ((BddNode) b2).right));
+                        bddDiff(bddUnion(((BddNode) b1).left, ((BddNode) b1).middle),
```

Comment:
@rdhananjaya should we use bddNode local variables here (and in other bdd common ops) to avoid multiple casts and improve readability?

Reply:
+1, can we all them `bdd1`, `bdd2`